cpp/src/plasma/client.cc

```diff
@@ -808,20 +808,27 @@ Status PlasmaClient::Impl::Abort(const ObjectID& object_id) {
   return ReadAbortReply(buffer.data(), buffer.size(), &id);
 }
 
-Status PlasmaClient::Impl::Delete(const ObjectID& object_id) {
+Status PlasmaClient::Impl::Delete(const std::vector<ObjectID>& object_ids) {
   RETURN_NOT_OK(FlushReleaseHistory());
-  // If the object is in used, client can't send the remove message.
-  if (objects_in_use_.count(object_id) > 0) {
-    return Status::UnknownError("PlasmaClient::Object is in use.");
-  } else {
-    // If we don't already have a reference to the object, we can try to remove the object
-    RETURN_NOT_OK(SendDeleteRequest(store_conn_, object_id));
+  std::vector<ObjectID> not_in_use_ids;
+  for (auto& object_id : object_ids) {
+    // If the object is in used, skip it.
+    if (objects_in_use_.count(object_id) == 0) {
+      not_in_use_ids.push_back(object_id);
+    }
+  }
+  if (not_in_use_ids.size() > 0) {
+    RETURN_NOT_OK(SendDeleteRequest(store_conn_, not_in_use_ids));
     std::vector<uint8_t> buffer;
     RETURN_NOT_OK(PlasmaReceive(store_conn_, MessageType::PlasmaDeleteReply, &buffer));
-    ObjectID object_id2;
     DCHECK_GT(buffer.size(), 0);
-    RETURN_NOT_OK(ReadDeleteReply(buffer.data(), buffer.size(), &object_id2));
+    std::vector<PlasmaError> error_codes;
+    not_in_use_ids.clear();
+    RETURN_NOT_OK(
+        ReadDeleteReply(buffer.data(), buffer.size(), not_in_use_ids, error_codes));
     return Status::OK();
+  } else {
+    return Status::UnknownError("PlasmaClient::All Objects are in use.");
```

Comment:
Hmm, let's make the semantics of this method "Delete all objects that are not in use", and get rid of this error, since it is confusing (why only raise the error if _all_ objects are in use as opposed to if _some_ objects are in use)?

Actually the best semantics in my opinion is this: If objects are in use, send a delete request to the object store anyways and it will clean the object as soon as no client is using it any more. Are you interested in implementing that? In any case we should open a JIRA to do it in the future and/or have more discussion!

Reply:
I did this change because I don't want to break the unit-test for deleting single object RPC and add unit-test for the new RPC. However, I accept that it really looks wired. I will remove this error as well as PlasmaError::ObjectNonexistent in protocol.cc.

Yes, i'm interested in implementing that. You can create a JIRA issue and assign it to me. In my understanding, we could add a cache to keep the objects to be deleted. Once it is not in use, we can clean it and remove it from the cache.